### PR TITLE
Parse 3DS Liability Shift in Card Approve Order Result

### PR DESCRIPTION
**Pull request by sshropshire**
### Summary of changes

 - Parse `liability_shift` parameter from `confirm-payment-source` 3DS contingency deep link response
 - Parse `error` parameter from `confirm-payment-source` 3DS contingency deep link response
 - Return `PayPalSDKError` when an error is detected in the 3DS contingency deep link response

 ### Checklist

 - [x] Added a changelog entry

- @sshropshire

### CODE REVIEW

File: CardPayments/src/main/java/com/paypal/android/cardpayments/CardClient.kt [hunk 1]
```diff
@@ -160,26 +160,35 @@ class CardClient internal constructor(
...
-                    val result = CardResult(metadata.orderId, deepLinkUrl)
-                    notifyApproveOrderSuccess(result)
-                } catch (error: PayPalSDKError) {
-                    analyticsService.sendAnalyticsEvent(
-                        "card-payments:3ds:get-order-info:failed",
-                        metadata.orderId
-                    )
-                    throw error
-                }
+            try {
```
Comment (sshropshire) [inline CardPayments/src/main/java/com/paypal/android/cardpayments/CardClient.kt:163]: This internals of this method can work without coroutines.

File: CardPayments/src/main/java/com/paypal/android/cardpayments/CardClient.kt [hunk 2]
```diff
@@ -160,26 +160,35 @@ class CardClient internal constructor(
...
+                analyticsService.sendAnalyticsEvent(
+                    "card-payments:3ds:get-order-info:succeeded",
+                    metadata.orderId
+                )
```
Comment (scannillo) [inline CardPayments/src/main/java/com/paypal/android/cardpayments/CardClient.kt]: Looks like when we removed the getOrder API call (after we had to move from accessTokens to clientIDs) we never removed these analytic events. Could be another PR, just calling it out.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,6 +1,14 @@
 # PayPal Android SDK Release Notes
 
+## unreleased
+
+* CardPayments
+  * Add `liabilityShift` property to `CardResult`
+  * Parse `liabilityShift` from deep link return URL in `CardClient#approveOrder()` browser-switched flow
+  * Callback `PayPalSDKError` when CardClient#approveOrder()` 3DS verification fails
```
Comment (scannillo) [inline CHANGELOG.md]:
```suggestion
  * Callback `PayPalSDKError` when `CardClient#approveOrder()` 3DS verification fails
```

**Review by scannillo — APPROVED** [review]

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,6 +1,14 @@
 # PayPal Android SDK Release Notes
 
+## unreleased
+
+* CardPayments
+  * Add `liabilityShift` property to `CardResult`
+  * Parse `liabilityShift` from deep link return URL in `CardClient#approveOrder()` browser-switched flow
```
Comment (sarahkoop) [inline CHANGELOG.md]: This feels like an implementation detail - is it different from the line above (from a merchant perspective)?

File: CardPayments/src/main/java/com/paypal/android/cardpayments/CardClient.kt [hunk 5]
```diff
@@ -160,26 +160,35 @@ class CardClient internal constructor(
...
+                notifyApproveOrderFailure(error)
             }
         }
     }
 
+    @Throws(PayPalSDKError::class)
+    private fun parseApproveOrderDeepLink(orderId: String, deepLinkUrl: Uri?): CardResult {
+        val error = deepLinkUrl?.getQueryParameter("error")
```
Comment (sarahkoop) [inline CardPayments/src/main/java/com/paypal/android/cardpayments/CardClient.kt]: Should we also return an error if the deepLinkUrl is null? Looks like currently that scenario will just return a CardResult with null params - not sure what the desired behavior is

File: CardPayments/src/main/java/com/paypal/android/cardpayments/CardClient.kt (hunk 2 above)
Comment (sshropshire) [reply CardPayments/src/main/java/com/paypal/android/cardpayments/CardClient.kt]: Oh good call. I'll go ahead and remove 👍 .

File: CHANGELOG.md (hunk 4 above)
Comment (sshropshire) [reply CHANGELOG.md]: That's true. I'll remove this too.

File: CardPayments/src/main/java/com/paypal/android/cardpayments/CardClient.kt (hunk 5 above)
Comment (sshropshire) [reply CardPayments/src/main/java/com/paypal/android/cardpayments/CardClient.kt]: Good catch I'll throw an error here when that happens. A user could technically clear their storage whenever they want.

Comment (sshropshire) [reply CardPayments/src/main/java/com/paypal/android/cardpayments/CardClient.kt]: Actually thinking it through this won't occur on a storage clear. Either way added a test case for it 👍 .

**Review by sarahkoop — APPROVED** [review]

File: CardPayments/src/main/java/com/paypal/android/cardpayments/CardClient.kt [hunk 6]
```diff
@@ -160,26 +160,35 @@ class CardClient internal constructor(
...
+            throw CardError.threeDSVerificationError
+        }
+
+        val error = deepLinkUrl.getQueryParameter("error")
+        if (error != null) {
+            throw CardError.threeDSVerificationError
+        } else {
+            val liabilityShift = deepLinkUrl.getQueryParameter("liability_shift")
+            return CardResult(orderId, deepLinkUrl, liabilityShift)
+        }
```
Comment (KunJeongPark) [inline CardPayments/src/main/java/com/paypal/android/cardpayments/CardClient.kt:181]: Steven, in iOS, we also check for conditions for success and throw another error if it is not in the format specified. To me, it seems risky to assume success if error parameter is absent.
https://github.com/paypal/paypal-ios/blob/2896471c1533de72880afd18545a33a374f9cfc0/Sources/CardPayments/CardClient.swift#L140
What do you think?

Comment (sshropshire) [reply CardPayments/src/main/java/com/paypal/android/cardpayments/CardClient.kt:181]: I can add those checks in 👍 . I missed that part from the iOS PR.

Comment (sshropshire) [reply CardPayments/src/main/java/com/paypal/android/cardpayments/CardClient.kt:181]: Added here in [af7cbad](https://github.com/paypal/paypal-android/pull/227/commits/af7cbad6aede5a940ae13cc34ed4fd1bc06bb05f).
